### PR TITLE
sql: add a framework for lazy generation of virtual table rows

### PR DESCRIPTION
**Pull request by rohany**
This PR adds a framework for lazily generating rows for virtual
tables. It also ports some virtual table generators that loop
over all table descriptors to use this framework.

Release note: None

### CODE REVIEW

**Comment by rohany** [comment]
Round 2 version of #46937.

This version looks significantly easier to use (i had to change only 3-4 LOC to convert these two into generators) and looks composable. These two already probably make it a better choice than #46937.

I'll wait on an LGTM before converting more tables over.

File: pkg/sql/crdb_internal.go
Comment (otan) [inline pkg/sql/crdb_internal.go:1679]: is there a reason we can't just return the `worker`, and let this boilerplate bit be somewhere else?

Comment (rohany) [reply pkg/sql/crdb_internal.go:1679]: There are some existing generators that don't necessarily fit into this pattern, so I don't want to push it up into the virtual table infrastructure yet.

File: pkg/sql/virtual_table.go
Comment (otan) [inline pkg/sql/virtual_table.go:50]: hmm, this  first select doesn't seem necessary. can you describe the race a bit more closely?

File: pkg/sql/virtual_table.go
Comment (otan) [inline pkg/sql/virtual_table.go:75]: can we close the send channel instead to signify completion?

**Comment by rohany** [comment]
hmm, this might be a bust because you aren't allowed to use txn's concurrently... I'll see if there is a way around this.

File: pkg/sql/virtual_table.go
Comment (rohany) [reply pkg/sql/virtual_table.go]: Yeah i don't think it is necessary, it was to catch people calling next after cleanup which is illegal anyway.

File: pkg/sql/virtual_table.go
Comment (rohany) [reply pkg/sql/virtual_table.go:88]: done

**Comment by rohany** [comment]
PTAL, added some extra sync to avoid concurrent transaction operations by the generators.

File: pkg/sql/virtual_table_test.go
Comment (otan) [inline pkg/sql/virtual_table_test.go:46]: super nit: seems like we have two separate tests here. you can spread these into two sub tests with `t.Run` (or split into two tests)

File: pkg/sql/virtual_table.go
Comment (otan) [inline pkg/sql/virtual_table.go:49]: i think we should pass ctx in here for <-ctx.Done() as well

File: pkg/sql/virtual_table.go
Comment (otan) [inline pkg/sql/virtual_table.go:33]: i would remove the "predefined error" bit in cast the user gets this and gets confused

File: pkg/sql/virtual_table.go
Comment (otan) [inline pkg/sql/virtual_table.go:70]: could just inline say `go func() {` instead of the startGenerator declaration if you wanted to :P

File: pkg/sql/crdb_internal.go
Comment (otan) [inline pkg/sql/crdb_internal.go:3013]: another nit from me: can we arrange the arguments to be the same order as the return

File: pkg/sql/crdb_internal.go
Comment (otan) [inline pkg/sql/crdb_internal.go:252]: i'm wondering if this is dangerous. do rows get copied else? can we add documentation that "assumes rows from addRow have been copied out and is safe to reuse"

**Review by otan — COMMENTED** [review]
i'd hold off on converting the rest into a separate PR, since this one would get huge already.

File: pkg/sql/crdb_internal.go
Comment (rohany) [reply pkg/sql/crdb_internal.go:252]: They get copied as part of the next interface if needed, but it makes sense to add this to the setupGenereator docs.

File: pkg/sql/crdb_internal.go
Comment (rohany) [reply pkg/sql/crdb_internal.go]: done

File: pkg/sql/virtual_table.go
Comment (rohany) [reply pkg/sql/virtual_table.go]: done

File: pkg/sql/virtual_table.go
Comment (rohany) [reply pkg/sql/virtual_table.go]: Done

File: pkg/sql/virtual_table_test.go
Comment (rohany) [reply pkg/sql/virtual_table_test.go]: Done

**Comment by rohany** [comment]
> i'd hold off on converting the rest into a separate PR, since this one would get huge already. 

Ok, makes sense.

File: pkg/sql/virtual_table.go
Comment (rohany) [reply pkg/sql/virtual_table.go:51]: I'm not sure if this is ok -- I don't know if the context passed into generate is the same one that can be canceled when we start making calls to next.

Comment (otan) [reply pkg/sql/virtual_table.go:51]: hmm, maybe next or addRow should take in a context instead?

**Review by jordanlewis — COMMENTED** [review]
Really cool stuff

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-:-M508gzc9PGdy5UTjjR7:be478ml)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @rohany)

___
*[pkg/sql/crdb_internal.go, line 1683 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M507djv84twEmMfxXq1:-M507djw1DhxNAeNOocZ:b-5dkr1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/crdb_internal.go#L1683)):*
> ```Go
> 							tree.MakeDBool(idx.Type == sqlbase.IndexDescriptor_INVERTED),
> 						)
> 						if err := addRow(row...); err != nil {
> ```

This is cool. I was going to complain that it didn't make sense, but I forgot that addRow actually blocks this method until the row has been sucked away / copied... I think that this is surprising given the name of `addRow`.

Can we change it to `pushRow`?  I think that will be slightly more obvious, but I also think that even still it suffers from underdocumentation because of the heavy use of closures.

What if instead of taking `pushRow` as an argument, this actually took an interface as an argument that included a well-defined `push` method? That would kind of unify its terminology with the `RowReceiver` interface in `pkg/sql/execinfra/base.go` that's used in DistSQL.
___
*[pkg/sql/virtual_table.go, line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M4ofvrylD7F-1lDu58T-r3-49:-M506xZv64KqmfL0hTDa:b-5mkixs) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fb1c22a8940bf9de60a29d3a0cc625ee0891d1f/pkg/sql/virtual_table.go#L49)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

hmm, maybe next or addRow should take in a context instead?
</blockquote></details>

I think you should use a context too. I forgot when we were talking about cancellation messages in SQL. Cancelling a query cancels the query's context, and long running operators are supposed to pay attention to that if possible. I also think that server shut down hooks into contexts, and ignoring context cancellation might be problematic.
___
*[pkg/sql/virtual_table.go, line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M506bNMA10q7OpaILfw:-M506bNN5qCb7AYLKlIM:b6bctu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/virtual_table.go#L36)):*
> ```Go
> 
> // setupGenerator takes in a worker that generates rows eagerly and transforms
> // it into a lazy row generator. It returns three functions:
> ```

s/three/two/?
___
*[pkg/sql/virtual_table.go, line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M506RAjCPccYbjqNXUJ:-M506RAjCPccYbjqNXUK:bvpdfoa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/virtual_table.go#L44)):*
> ```Go
> func setupGenerator(
> 	worker func(addRow func(...tree.Datum) error) error,
> ) (next virtualTableGenerator, cleanup func()) {
> ```

Define a type for func(), so that it's obvious what return values are.

Like:

```
type CleanupFunc func()
```
___
*[pkg/sql/virtual_table_test.go, line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M507MmU99_IO3x5BNCm:-M507MmU99_IO3x5BNCn:b-d98qkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/virtual_table_test.go#L74)):*
> ```Go
> 		_, cleanup := setupGenerator(worker)
> 		cleanup()
> 	})
> ```

Do a quick benchmark?

**Comment by rohany** [comment]
> I think you should use a context too. I forgot when we were talking about cancellation messages in SQL. Cancelling a query cancels the query's context, and long running operators are supposed to pay attention to that if possible. I also think that server shut down hooks into contexts, and ignoring context cancellation might be problematic.

I'm looking to using a context for this as well, but i'm not sure which one -- is it right to use the context that is passed into the setup for the virtual table generator? Based on this generator approach it doesn't seem feasible to get the context that is passed into calls to next over to the generator.

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47440)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @otan)

___
*[pkg/sql/crdb_internal.go, line 1683 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M507djv84twEmMfxXq1:-M52ahJLDqs6nfdn3kEp:b-ptppxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/crdb_internal.go#L1683)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This is cool. I was going to complain that it didn't make sense, but I forgot that addRow actually blocks this method until the row has been sucked away / copied... I think that this is surprising given the name of `addRow`.

Can we change it to `pushRow`?  I think that will be slightly more obvious, but I also think that even still it suffers from underdocumentation because of the heavy use of closures.

What if instead of taking `pushRow` as an argument, this actually took an interface as an argument that included a well-defined `push` method? That would kind of unify its terminology with the `RowReceiver` interface in `pkg/sql/execinfra/base.go` that's used in DistSQL.
</blockquote></details>

Done, it makes this clearer.
___
*[pkg/sql/virtual_table.go, line 49 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M4ofvrylD7F-1lDu58T-r3-49:-M52amKpBYLsooQJh7O7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7fb1c22a8940bf9de60a29d3a0cc625ee0891d1f/pkg/sql/virtual_table.go#L49)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think you should use a context too. I forgot when we were talking about cancellation messages in SQL. Cancelling a query cancels the query's context, and long running operators are supposed to pay attention to that if possible. I also think that server shut down hooks into contexts, and ignoring context cancellation might be problematic.
</blockquote></details>

Done.
___
*[pkg/sql/virtual_table.go, line 36 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M506bNMA10q7OpaILfw:-M52alCg65YeIyGilRAQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/virtual_table.go#L36)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

s/three/two/?
</blockquote></details>

Done.
___
*[pkg/sql/virtual_table.go, line 44 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M506RAjCPccYbjqNXUJ:-M52alSqCm_OxIVnXHQS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/virtual_table.go#L44)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Define a type for func(), so that it's obvious what return values are.

Like:

```
type CleanupFunc func()
```
</blockquote></details>

Done.
___
*[pkg/sql/virtual_table_test.go, line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M507MmU99_IO3x5BNCm:-M52anV00yKEqRV7K5dS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/virtual_table_test.go#L74)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Do a quick benchmark?
</blockquote></details>

Done.

**Review by jordanlewis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47440)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rohany)

___
*[pkg/sql/virtual_table.go, line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5DHPWw3NFmPIINBcvq:-M5DHPWw3NFmPIINBcvr:b-9vsamn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L35)):*
> ```Go
> type rowPusher interface {
> 	// pushRow pushes the input row to the receiver of the generator. It does
> 	// mutate the input row. It will block until the the data has been received
> ```

What do you mean by "it does mutate the input row"? 

You should add the guarantees it exposes as well - like, once `pushRow` returns, the input container can be resized / overwritten safely.
___
*[pkg/sql/virtual_table.go, line 58 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5DHh0D36MFof_eBEyA:-M5DHh0D36MFof_eBEyB:b79o6xa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L58)):*
> ```Go
> // * next: A handle that can be called to generate a row from the worker. Next
> //   cannot be called once cleanup has been called.
> // * cleanup: Performs all cleanup. May only be called once.
> ```

This has to be called, right? You should write something like "the caller is responsible to call cleanup exactly once after this returns, or risk leaking resources."
___
*[pkg/sql/virtual_table.go, line 78 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J3uXXD59IR62TEX85:-M5J3uXXD59IR62TEX86:b-j9130l) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L78)):*
> ```Go
> 		}
> 
> 		select {
> ```

I think you need to wait on `ctx.Done()` in this select as well, or else server shutdown might hang...
___
*[pkg/sql/virtual_table.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J3x2R1Hssi3U9MjiI:-M5J3x2R1Hssi3U9MjiJ:b-jy3gwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L92)):*
> ```Go
> 		// which effectively gives ownership of the transaction usage over to the
> 		// worker, and then back to the next() caller after it is done.
> 		select {
> ```

Same with this one.
___
*[pkg/sql/virtual_table.go, line 106 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J4MrE6bThG5M_XXtL:-M5J4MrE6bThG5M_XXtM:btm236j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L106)):*
> ```Go
> 		// avoid leaking this goroutine.
> 		select {
> 		case <-done:
> ```

Have to wait on ctx.Done here too.
___
*[pkg/sql/virtual_table.go, line 123 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J4Q6qBCJGjFkox0WG:-M5J4Q6qBCJGjFkox0WH:b-kkd7vy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L123)):*
> ```Go
> 
> 	next = func() (tree.Datums, error) {
> 		allowWork <- struct{}{}
> ```

I think that you need to make these selects and include ctx.Done too, in case the push blocks and someone wants to turn off the server or cancel the query.
___
*[pkg/sql/virtual_table_test.go, line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M507MmU99_IO3x5BNCm:-M5J4XUzA7m05U0riVBV:bvpipm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/virtual_table_test.go#L74)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Results?

**Review by rohany — COMMENTED** [review]
Thanks for the catches -- this stuff is super tricky.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-:-M5JNwb_5PNfO-TsthAV:b7x00zu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @otan)

___
*[pkg/sql/virtual_table.go, line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5DHPWw3NFmPIINBcvq:-M5JNtjk6NO6rev16AaJ:bb0j96a) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L35)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

What do you mean by "it does mutate the input row"? 

You should add the guarantees it exposes as well - like, once `pushRow` returns, the input container can be resized / overwritten safely.
</blockquote></details>

Sorry typo -- updated!
___
*[pkg/sql/virtual_table.go, line 58 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5DHh0D36MFof_eBEyA:-M5JNiof43RCTWQKbYP1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L58)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This has to be called, right? You should write something like "the caller is responsible to call cleanup exactly once after this returns, or risk leaking resources."
</blockquote></details>

Done.
___
*[pkg/sql/virtual_table.go, line 78 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J3uXXD59IR62TEX85:-M5JNQeZDt3ByL24CPEM:b2gwt18) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L78)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think you need to wait on `ctx.Done()` in this select as well, or else server shutdown might hang...
</blockquote></details>

I don't think this one needs a check -- this one doesn't block. Also, i need to pull the check on ctx out to ensure the tests are deterministic in returning a querycancellederror
___
*[pkg/sql/virtual_table.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J3x2R1Hssi3U9MjiI:-M5JNPY317RXkwzrwqSQ:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L92)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Same with this one.
</blockquote></details>

done
___
*[pkg/sql/virtual_table.go, line 106 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J4MrE6bThG5M_XXtL:-M5JNCMR2KHPbcqDX3Ww:b-z8esyt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L106)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Have to wait on ctx.Done here too.
</blockquote></details>

Done -- i was handling this with the select in add row, but we should try and exit here too.
___
*[pkg/sql/virtual_table.go, line 123 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J4Q6qBCJGjFkox0WG:-M5JN8g-6Pq-xADcYTXH:b6078yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L123)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think that you need to make these selects and include ctx.Done too, in case the push blocks and someone wants to turn off the server or cancel the query.
</blockquote></details>

Good catch, added.
___
*[pkg/sql/virtual_table_test.go, line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M507MmU99_IO3x5BNCm:-M5JIWxe8txlPfNYtfyd:b-t5flbe) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/virtual_table_test.go#L74)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Results?
</blockquote></details>

Sorry, i definitely thought i pasted the benchmark results here...
```
pkg: github.com/cockroachdb/cockroach/pkg/sql
BenchmarkVirtualTableGenerators/bench_read-16         	  394672	      2796 ns/op
PASS
```
So it takes about 3 ms to stream 1k rows through.

**Review by jordanlewis — COMMENTED** [review]
NP, yeah it's definitely tricky. Maybe it would be good to add a higher-level test? Looks like there are some cancellation tests in `run_control_test.go`. Besides that I think this :lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-:-M5J_3Ir90xOd0u8GilX:bgqk48x)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, and @rohany)

___
*[pkg/sql/virtual_table.go, line 38 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J_tb69q-Xz7HetGnS:-M5J_tb69q-Xz7HetGnT:bt32qvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/ccb0285d5d22ab80e32493bf3bf401448a445585/pkg/sql/virtual_table.go#L38)):*
> ```Go
> 	// and more data has been requested. Once pushRow returns, the caller is free
> 	// to mutate the slice passed as input. The caller is not allowed to perform
> 	// operations on a transaction while blocked on a call pushRow.
> ```

I think you're missing a "to" in between call and pushRow now :)
___
*[pkg/sql/virtual_table_test.go, line 74 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M507MmU99_IO3x5BNCm:-M5J_RNf-JvDzxvzxXOS:b-qujygn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ff632fd1db652ce6fb11a37a1f3e4a8840a15a4/pkg/sql/virtual_table_test.go#L74)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Sorry, i definitely thought i pasted the benchmark results here...
```
pkg: github.com/cockroachdb/cockroach/pkg/sql
BenchmarkVirtualTableGenerators/bench_read-16         	  394672	      2796 ns/op
PASS
```
So it takes about 3 ms to stream 1k rows through.
</blockquote></details>

Ok, that seems good enough to me.

**Comment by rohany** [comment]
Fixed the typos.

I tried adding a test like the ones in `run_control_test.go` but ran into #47684. Either way, it was tough to catch the query before it finished without doing something absurd like creating 1000 tables. Otherwise it would actually cancel the query in time only a small percentage of the time.

**Review by jordanlewis — COMMENTED** [review]
Ok, fair enough - I guess you could do a testing knob but it might not be worth it.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-:-M5KkrZo2r1uuEQzmKd6:b-quk0w2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @rohany)

___
*[pkg/sql/virtual_table.go, line 78 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J3uXXD59IR62TEX85:-M5KkdSGDSopCNpgGiyS:b-fv2x8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L78)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I don't think this one needs a check -- this one doesn't block. Also, i need to pull the check on ctx out to ensure the tests are deterministic in returning a querycancellederror
</blockquote></details>

Couldn't it block if nothing was on `done` and `send` was full?
___
*[pkg/sql/virtual_table.go, line 66 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5Knb7O38vYfFWQULPR:-M5Knb7P7QKtIFqFnZ67:bee9tu5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0001335d4458f545284e71e4a4735307076182c/pkg/sql/virtual_table.go#L66)):*
> ```Go
> 	ctx context.Context, worker func(pusher rowPusher) error,
> ) (next virtualTableGenerator, cleanup cleanupFunc) {
> 	allowWork, done, send := make(chan struct{}, 1), make(chan struct{}), make(chan virtualTableGeneratorResponse, 1)
> ```

I think we can eliminate the `done` channel by making a child context:

```
ctx, cleanup = context.WithCancel(ctx)
```

Then, anywhere you're listening on both done and ctx.Done(), you can just listen to ctx.Done(). I think that will make this whole thing be a bit cleaner.

I also think we could get rid of `allowWork` completely by using the `send` channel both to send responses and to listen for wakeups - the worker goroutine would send to `send` and then immediately read from `send` and `ctx.Done()`, and `next` would do the same.

Also, in either case, I think you want `allowWork` and `send` to be unbuffered channels, right? A channel of size 1 actually lets you put something into it in a nonblocking way, and it'll block on the second send - but don't we want to block on the first send, until the remote end takes the object?

**Review by knz — COMMENTED** [review]
> Reviewed 1 of 4 files at r3, 3 of 4 files at r5, 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47440)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @otan, and @rohany)

___
*[pkg/sql/virtual_table.go, line 78 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-M5J3uXXD59IR62TEX85:-M5LrSwMD5gmrMOITecs:b-qkjo2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e46e9b7e34ba9cff7b172772d517aac13693eee7/pkg/sql/virtual_table.go#L78)):*
I agree with jordan the two selects need to be merged. As-is your code can misbehave upon server shutdown.

>  to ensure the tests are deterministic in returning a querycancellederror

if that is your concern, maybe the test needs to be adjusted.

**Comment by rohany** [comment]
That makes sense, I was able to remove the done channel with a child context. 

> I also think we could get rid of allowWork completely by using the send channel both to send responses and to listen for wakeups - the worker goroutine would send to send and then immediately read from send and ctx.Done(), and next would do the same.

We can't do this, because this opens a chance for the worker to write into send and then immediately read its own write.

Yeah, I already made both the channels buffered with size 1

**Comment by jordanlewis** [comment]
And if you made the channel unbuffered, then the worker would immediately
block on send - so it wouldn't have the read-own-write problem.

>

**Comment by rohany** [comment]
Alright, i think the one channel solution is good -- we're able to also wait on ctx.Done in every place where an operation may block, so we'll be good for cancellations.

**Review by jordanlewis — APPROVED** [review]
Yay, this looks clean now. :lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47440#-:-M5NOAwGC9DRIZ2-xBtW:b-ucll3l)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @otan)

**Comment by rohany** [comment]
bors r=jordanlewis

**Pull request merged**
